src/encoder.rs

```diff
@@ -136,7 +191,110 @@ impl<W: Write> Writer<W> {
+        match self.info {
+            Info { animation_control: Some(_), frame_control: Some(frame_control), ..} => {
+                if frame_control.sequence_number != 0 {
+                    Err(EncodingError::Format("separate default image provided after frame sequence has begun".into()))
+                } else if self.separate_default_image {
+                    Err(EncodingError::Format("default image already written".into()))
+                } else {
+                    self.separate_default_image = true;
+                    self.write_image_data(data)
+                }
+            }
+            _ => {
+                Err(EncodingError::Format("default image provided for a non-animated PNG".into()))
+            }
+        }
+    }
+
+    #[allow(non_snake_case)]
+    fn write_fcTL(&mut self) -> Result<()> {
+        let frame_ctl = self.info.frame_control.ok_or(EncodingError::Format("cannot write fcTL for a non-animated PNG".into()))?;
+        let mut data = [0u8; 26];
+
+        (&mut data[..]).write_be(frame_ctl.sequence_number)?;
+        (&mut data[4..]).write_be(frame_ctl.width)?;
+        (&mut data[8..]).write_be(frame_ctl.height)?;
+        (&mut data[12..]).write_be(frame_ctl.x_offset)?;
+        (&mut data[16..]).write_be(frame_ctl.y_offset)?;
+        (&mut data[20..]).write_be(frame_ctl.delay_num)?;
+        (&mut data[22..]).write_be(frame_ctl.delay_den)?;
+        data[24] = frame_ctl.dispose_op as u8;
+        data[25] = frame_ctl.blend_op as u8;
+
+        self.write_chunk(chunk::fcTL, &data)
+    }
+
+    #[allow(non_snake_case)]
+    fn write_fdAT(&mut self, data: &[u8]) -> Result<()> {
+        // println!("Writing fdAT:{:?}", self.info.frame_control.unwrap().sequence_number+1);
+
+        let zlib = self.get_image_data(data)?;
+
+        let mut data = [0u8; 4];
+        (&mut data[..]).write_be(self.info.frame_control
+                .ok_or(EncodingError::Format("cannot write fdAT for a non-animated PNG".into()))?.sequence_number+1u32)?;
+
+        self.write_chunk_with_fields(chunk::fdAT, &zlib.finish()?, Some(&data))
+    }
+
+    pub fn write_frame(&mut self, data: &[u8]) -> Result<()> {
+        // println!("{:?}", self.info.frame_control.unwrap().sequence_number);
+
+        match self.info {
+            Info { animation_control: Some(AnimationControl { num_frames: 0, ..}) , frame_control: Some(_), ..} => {
+                Err(EncodingError::Format("exceeded number of frames specified".into()))
+            },
+            Info { animation_control: Some(anim_ctl), frame_control: Some(frame_control), ..} => {
+                match frame_control.sequence_number {
+                    0 => {
+                        let ret = if self.separate_default_image { // If we've already written the default image we can write frames the normal way
+                            // fcTL + fdAT
+
+                            self.write_fcTL().and(self.write_fdAT(data))
```

Comment:
This kind of chaining could lead to data corruption as the `write_fdAT()` operation will be executed regardless of whether or not `write_fcTL()` succeeds. These should instead be wrapped with `try!()` like all the other fallible operations (although the `?` operator has been stable for a long time now, `try!()` would be consistent with the rest of the crate).